File: 2nd Year/2nd Semester/MPP/Triatlon-C#/TriatlonNetworking/jsonprotocol/Response.cs
using System.Text;
using TriatlonNetworking.dto;

namespace TriatlonNetworking.jsonprotocol;

[Serializable]
public class Response
{
    public ResponseType Type { get; set; }
    public string ErrorMessage { get; set; }

    // Individual DTOs
    public ArbitruDTO ArbitruDTO { get; set; }
    public ParticipantDTO ParticipantDTO { get; set; }
    public ProbaDTO ProbaDTO { get; set; }
    public RezultatDTO RezultatDTO { get; set; }

    // Arrays of DTOs
    public ArbitruDTO[] ArbitriiDTO { get; set; }
    public ParticipantDTO[] ParticipantiDTO { get; set; }
    public ProbaDTO[] ProbeDTO { get; set; }
    public RezultatDTO[] RezultateDTO { get; set; }

    public Response()
    {
    }

    public TipProbaDTO GetTipProba()
    {
        return ProbaDTO?.TipProba ?? default(TipProbaDTO);
    }

    public void SetTipProba(TipProbaDTO tipProba)
    {
        if (ProbaDTO != null)
        {
            ProbaDTO.TipProba = tipProba;
        }
    }

    public override string ToString()
    {
        var sb = new StringBuilder();
        sb.Append("Response{");
        sb.Append($"type={Type}, ");
        sb.Append($"errorMessage='{ErrorMessage}', ");
        sb.Append($"arbitruDTO={ArbitruDTO}, ");
        sb.Append(
            $"arbitriiDTO={string.Join(", ", (IEnumerable<ArbitruDTO>)(ArbitriiDTO ?? Array.Empty<ArbitruDTO>()))}, ");
        sb.Append($"participantDTO={ParticipantDTO}, ");
        sb.Append(
            $"participantiDTO={string.Join(", ", (IEnumerable<ParticipantDTO>)(ParticipantiDTO ?? Array.Empty<ParticipantDTO>()))}, ");
        sb.Append($"probaDTO={ProbaDTO}, ");
        sb.Append($"probeDTO={string.Join(", ", (IEnumerable<ProbaDTO>)(ProbeDTO ?? Array.Empty<ProbaDTO>()))}, ");
        sb.Append($"rezultatDTO={RezultatDTO}, ");
        sb.Append(
            $"rezultateDTO={string.Join(", ", (IEnumerable<RezultatDTO>)(RezultateDTO ?? Array.Empty<RezultatDTO>()))}");
        sb.Append("}");
        return sb.ToString();
    }
}

File: 2nd Year/2nd Semester/MPP/Triatlon-Java/TriatlonNetworking/src/main/java/triatlon/network/jsonprotocol/Response.java
package triatlon.network.jsonprotocol;

import triatlon.model.Participant;
import triatlon.network.dto.*;
import triatlon.model.Rezultat;

import java.io.Serializable;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

public class Response implements Serializable {
    private ResponseType type;
    private String errorMessage;

    // Individual DTOs
    private ArbitruDTO arbitruDTO;
    private ParticipantDTO participantDTO;
    private ProbaDTO probaDTO;
    private RezultatDTO rezultatDTO;

    // Arrays of DTOs
    private ArbitruDTO[] arbitriiDTO;
    private ParticipantDTO[] participantiDTO;
    private ProbaDTO[] probeDTO;
    private RezultatDTO[] rezultateDTO;

    public Response() {
    }

    // Getters and setters for type and error message
    public ResponseType getType() {
        return type;
    }

    public void setType(ResponseType type) {
        this.type = type;
    }

    public String getErrorMessage() {
        return errorMessage;
    }

    public void setErrorMessage(String errorMessage) {
        this.errorMessage = errorMessage;
    }

    // Getters and setters for individual DTOs
    public ArbitruDTO getArbitruDTO() {
        return arbitruDTO;
    }

    public void setArbitruDTO(ArbitruDTO arbitruDTO) {
        this.arbitruDTO = arbitruDTO;
    }

    public ParticipantDTO getParticipantDTO() {
        return participantDTO;
    }

    public void setParticipantDTO(ParticipantDTO participantDTO) {
        this.participantDTO = participantDTO;
    }

    public ProbaDTO getProbaDTO() {
        return probaDTO;
    }

    public void setProbaDTO(ProbaDTO probaDTO) {
        this.probaDTO = probaDTO;
    }

    public RezultatDTO getRezultatDTO() {
        return rezultatDTO;
    }

    public void setRezultatDTO(RezultatDTO rezultatDTO) {
        this.rezultatDTO = rezultatDTO;
    }

    // Getters and setters for arrays of DTOs
    public ArbitruDTO[] getArbitrii() {
        return arbitriiDTO;
    }

    public void setArbitrii(ArbitruDTO[] arbitriiDTO) {
        this.arbitriiDTO = arbitriiDTO;
    }

    public ParticipantDTO[] getParticipanti() {
        return participantiDTO;
    }

    public void setParticipanti(ParticipantDTO[] participantiDTO) {
        this.participantiDTO = participantiDTO;
    }

    public ProbaDTO[] getProbe() {
        return probeDTO;
    }

    public void setProbe(ProbaDTO[] probeDTO) {
        this.probeDTO = probeDTO;
    }

    public RezultatDTO[] getRezultateDTO() {
        return rezultateDTO;
    }

    public void setRezultateDTO(RezultatDTO[] rezultateDTO) {
        this.rezultateDTO = rezultateDTO;
    }

    // Conversion method that correctly uses DTOUtils
    public Rezultat[] getRezultate() {
        if (rezultateDTO == null) {
            return null;
        }
        return DTOUtils.getFromDTO(rezultateDTO);
    }

    @Override
    public String toString() {
        return "Response{" +
                "type=" + type +
                ", errorMessage='" + errorMessage + '\'' +
                ", arbitruDTO=" + arbitruDTO +
                ", arbitriiDTO=" + Arrays.toString(arbitriiDTO) +
                ", participantDTO=" + participantDTO +
                ", participantiDTO=" + Arrays.toString(participantiDTO) +
                ", probaDTO=" + probaDTO +
                ", probeDTO=" + Arrays.toString(probeDTO) +
                ", rezultatDTO=" + rezultatDTO +
                ", rezultateDTO=" + Arrays.toString(rezultateDTO) +
                '}';
    }

    public TipProbaDTO getTipProba() {
        if (probaDTO == null) {
            return null;
        }
        return probaDTO.getTipProba();
    }


    public void setTipProba(TipProbaDTO tipProba) {
        if (probaDTO != null) {
            probaDTO.setTipProba(tipProba);
        }
    }
}